### PR TITLE
Python SDK generation with pyproject.toml stops generating setup.py

### LINKED ISSUE #13709: [sdkgen/python] Building wheels with setup.py tries to install dev builds of plugins
When attempting to build a wheel with the following command:

```
python3 setup.py sdist bdist_wheel
```

It fails because `setup.py` contains some [code](https://github.com/pulumi/pulumi/blob/54033de4e64a9815cbf41c6e7b2dd3bf53675b73/pkg/codegen/testing/test/testdata/secrets/python/setup.py#L14-L29) that shells out to `pulumi plugin install`, and it's trying to install a dev build of a plugin that hasn't been published yet.

### Background

The purpose of shelling out to `pulumi plugin install` from `setup.py` is to have the Pulumi plugin installed at the same time that the package is installed. This works currently for source distributions published to PyPi, because `setup.py` will run when the package is installed from a source distribution.

However, when we start publishing wheels (in addition to source distributions), most users will be getting the wheel over the source distribution. Wheels have no mechanism to run arbitrary code when they are installed to be able to shell out to `pulumi plugin install`, which is going to make Python behave the same as all our other languages (except Node.js) w.r.t. plugin installation. Instead of the plugin being installed when the package installed, it'll be relying on the same mechanism the other languages rely on: before running a program, the engine asks the language host for plugins required by the program, and install those before running the program.

### Proposed Fixes

**Option 1**
I think the simplest fix is to make it so that if the flag to generate `pyproject.toml` is set, the SDK generator stops emitting the code that shells out to `pulumi plugin install` from `setup.py`. Then commands like `python3 setup.py sdist bdist_wheel` will succeed at build time. Installing from a source distribution will no longer have the `pulumi plugin install`, but that's the same behavior you'd get if installing the wheel.

**Option 2**
A potentially better solution would be to elide `setup.py` altogether if the flag is set. You'd then be more reliant on the `pyproject.toml` for building the source distribution and wheel, using a command like `python3 -m build` or `python3 --use-pep517 -m pip wheel .`. However, I believe this will require updates to the generated `pyproject.toml` to ensure files like `py.typed` and `pulumi-plugin.json` are included in the generated source distribution and wheel. We may also need to emit a `[build-system]` in the generated `pyproject.toml` per https://packaging.python.org/en/latest/tutorials/packaging-projects/#creating-pyproject-toml.

### PR DESCRIPTION
**Pull request by t0yv0**
This change affects projects that use Python SDK generator with the following option enabled:

    "pyproject": {"enable": true"}

Before this change, pyproject.toml was generated alongsie setup.py for such projects.

After the change, setup.py is no longer generated, by pyproject.toml is extended to opt into setuptools backend and ensure that building this project embeds metadata files:

    py.typed           # to support mypy and pyright users
    pulumi-plugin.json # to support Pulumi interop with the generated provider SDK

The generated code can then be built with the `build` module to produce source and/or wheel distributions. See also: https://pypa-build.readthedocs.io/en/latest/

# Description

Fixes #13709  - this takes Option 2 (drop setup.py, make pyproject.toml self-sufficient)

See also: https://setuptools.pypa.io/en/latest/userguide/datafiles.html#package-data

## Checklist

- [ ] I have run `make tidy` to update any new dependencies
- [ ] I have run `make lint` to verify my code passes the lint check
  - [ ] I have formatted my code using `gofumpt`

- [ ] I have added tests that prove my fix is effective or that my feature works

- [ ] I have run `make changelog` and committed the `changelog/pending/<file>` documenting my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Cloud API version

### CODE REVIEW

**Comment by t0yv0** [comment]
This approach is being pre-tested end to end in pulumi-kubernetes through make scripts 

https://github.com/pulumi/pulumi-kubernetes/pull/2544
and
https://github.com/pulumi/home/issues/2883

for context.

**Review by justinvp — APPROVED** [review]
LGTM. Thanks, @t0yv0!

File: changelog/pending/20230827--sdkgen-nodejs-python--nodejs-and-python-packages-no-longer-running-plugin-install-scripts-on-package-install.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+changes:
+- type: feat
+  scope: sdkgen/nodejs,python
+  description: NodeJS and Python packages no longer running plugin install scripts on package install.
```
[the remark this answers is not in the document]
Comment (t0yv0) [reply changelog/pending/20230827--sdkgen-nodejs-python--nodejs-and-python-packages-no-longer-running-plugin-install-scripts-on-package-install.yaml]: Yes this got removed after a rebase it looks like we're good. Thank you!

**Comment by t0yv0** [comment]
bors merge

**Pull request merged**